components/raftstore/src/store/fsm/peer.rs

```diff
@@ -3214,40 +3224,49 @@ where
             .peer
             .mut_store()
             .maybe_gc_cache(alive_cache_idx, applied_idx);
+
         let first_idx = self.fsm.peer.get_store().first_index();
-        let mut compact_idx;
-        if applied_idx > first_idx
-            && applied_idx - first_idx >= self.ctx.cfg.raft_log_gc_count_limit
-        {
-            compact_idx = applied_idx;
-        } else if self.fsm.peer.raft_log_size_hint >= self.ctx.cfg.raft_log_gc_size_limit.0 {
-            compact_idx = applied_idx;
-        } else if replicated_idx < first_idx
-            || replicated_idx - first_idx <= self.ctx.cfg.raft_log_gc_threshold
-        {
+
+        if !force_compact && replicated_idx < first_idx {
             return;
-        } else {
-            compact_idx = replicated_idx;
         }
 
+        if !force_compact && replicated_idx - first_idx < self.ctx.cfg.raft_log_gc_threshold {
```

Comment:
If there is only one raft entry left that is COMPACT_LOG type, also will trigger raft log gc after 6 rounds checking.